server/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java

```diff
@@ -654,34 +650,24 @@ public void testPrimaryPhaseExecutesDelegatedRequestOnRelocationTarget() throws
         AtomicBoolean executed = new AtomicBoolean();
         final TransportReplicationAction.ConcreteShardRequest<Request> primaryRequest
             = new TransportReplicationAction.ConcreteShardRequest<>(request, primaryShard.allocationId().getRelocationId(), primaryTerm);
-        action.new AsyncPrimaryAction(primaryRequest, listener, task) {
-            @Override
-            protected ReplicationOperation<Request, Request, TransportReplicationAction.PrimaryResult<Request, TestResponse>>
-            createReplicatedOperation(
-                    Request request,
-                    ActionListener<TransportReplicationAction.PrimaryResult<Request, TestResponse>> actionListener,
-                    TransportReplicationAction<Request, Request, TestResponse>.PrimaryShardReference primaryShardReference) {
-                return new NoopReplicationOperation(request, actionListener, primaryTerm) {
-                    @Override
-                    public void execute() throws Exception {
-                        assertPhase(task, "primary");
-                        assertFalse(executed.getAndSet(true));
-                        super.execute();
-                    }
-                };
-            }
 
+        new TestAction(Settings.EMPTY, "internal:testAction2", transportService, clusterService, shardStateAction, threadPool) {
             @Override
-            public void onFailure(Exception e) {
-                throw new RuntimeException(e);
+            protected void shardOperationOnPrimary(Request shardRequest, IndexShard primary,
+                                                   ActionListener<PrimaryResult<Request, TestResponse>> listener) {
+                assertPhase(task, "primary");
+                assertFalse(executed.getAndSet(true));
+                super.shardOperationOnPrimary(shardRequest, primary, listener);
             }
-        }.run();
+        }.new AsyncPrimaryAction(primaryRequest, ActionListener.wrap(listener::onResponse, throwable -> {
```

Comment:
Could we instead of using `ActionListener.wrap` just assert that `listener.isDone()` and do `listener.get()` like in the test above?

Reply:
Yes, it seems we can. I pushed 1350c0fb058be3e38460643dee492c85b4d63316.